File: coding-patterns/16-binary-search/121.py
# Bitonic Array Maximum (Easy)

class Solution:
  def findMax(self, arr):
    start, end = 0, len(arr) - 1
    while start < end:
      mid = start + (end - start) // 2
      if arr[mid] > arr[mid + 1]:
        end = mid
      else:
        start = mid + 1

    # at the end of the while loop, 'start == end'
    return arr[start]

File: coding-patterns/16-binary-search/121.cpp
// Bitonic Array Maximum (Easy)

using namespace std;

#include <iostream>
#include <vector>

class Solution {
public:
  int findMax(const vector<int>& arr) {
    int start = 0;
    int end = arr.size() - 1;
    while (start < end) {
        int mid = start + (end - start) / 2;
        if (arr[mid] > arr[mid + 1])
            end = mid;
        else
            start = mid + 1;
    }
    return arr[start];
  }
};
